server/src/server.ts

```diff
@@ -918,71 +873,237 @@ function getQuickFix(diagnostic: Diagnostic, title: string, range: Range, replac
+			continue;
+		}
+		const FUNC = "func";
+		const STRUCT = "struct";
+		const NAMESPACE = "namespace";
+		const isFunction = line.startsWith(FUNC) && line.length > FUNC.length+1 && line.charAt(FUNC.length).match(/\s/);
+		const isStruct = line.startsWith(STRUCT) && line.length > STRUCT.length+1 && line.charAt(STRUCT.length).match(/\s/);
+		const isNamespace = line.startsWith(NAMESPACE) && line.length > NAMESPACE.length+1 && line.charAt(NAMESPACE.length).match(/\s/);
+		if (isFunction || isStruct || isNamespace) {
+			let importItem = line.split(/[\s{(:]+/)[1];
+			importItems.add(importItem);
+		}				
+	}
+
+	return importItems;
+}
+
+/**
+ * @summary Gets text before and after the cursor ( on the same line )
+ * @param position 
+ * @param textDocumentFromURI 
+ * @returns Text before and after the cursor position | undefined
+ */
+function getTextAroundCursor(position: Position, textDocumentFromURI: TextDocument) {
+	const startOfLine = { 
+		line: position.line, 
+		character: 0 
+	};
+	const cursorPosition = { 
+		line: position.line, 
+		character: position.character
+	};
+	const nextLine = {
+		line: position.line + 1,
+		character: 0,
+	};
+
+	const lineUpToCursor = textDocumentFromURI.getText({ start: startOfLine, end: cursorPosition });
+	const lineUpToCursorSplit = lineUpToCursor.split(/\s+/);
+	if(lineUpToCursorSplit?.[0] === undefined)
+		return undefined
+
+	const textBeforeCursor = lineUpToCursorSplit[lineUpToCursorSplit.length - 1];
+		
+	const textAfterCursor = textDocumentFromURI
+		.getText({ start: cursorPosition, end: nextLine })?.split(/\s+/)[0];
+	
+	return { textBeforeCursor, textAfterCursor };
+}
+
+/**
+ * @summary Infers the type of the syntax the cursor is pointing at
+ * @param position 
+ * @param textDocumentFromURI 
+ * @returns SyntaxType of the current location
+ */
+function getSyntaxType(position: Position, textDocumentFromURI: TextDocument): SyntaxType {
+	const fileStart = { line: 0, character: 0 };
+	const cursorPosition = { line: position.line, character: position.character };
+
+	// Get text from the start of the document to position of the cursor
+	// This might be too exhaustive on a large file, should change if there's a better way to 
+	// determine if we're inside a function/from statement etc.
+	const textUpToCursor = textDocumentFromURI.getText({start: fileStart, end: cursorPosition});
+	
+	// The order here might matter, as they are evaluated top-to-bottom
+	const regexes = [
+		// func ...{}()
+		// Match func only if there's no end afterwards
+		{ exp: /^func(?!(.|\s)*^end)/gm, syn: SyntaxType.Function },
+		// from module import (
+		// Match only if there's no matching ")" meaning we're inside the parantheses
+		// After the parantheses it will also match any number of function names
+		{ exp: /^from[ \t]+[a-zA-Z0-9_.]*[ \t]+import[ \t]*\((?:[ \n\t]*[a-zA-Z0-9_.]*[ \t]*,*)*(?![\s\S]*\))/gm, syn: SyntaxType.ImportFunctionP },
```

Comment:
There seems to be problem with this regex.  When pasting this regex on [regex101.com](https://regex101.com/) and test string using [this](https://github.com/OpenZeppelin/cairo-contracts/blob/dfafdd2ebd778718cf440555e975f6eace278695/src/openzeppelin/token/erc721/ERC721_Mintable_Burnable.cairo), it shows "Catastrophic backtracking has been detected"

The language server seems to hang when doing code completion in various parts of that cairo file, which might be caused by this.

Reply:
I believe this happens because of the way I wrote the RegEx, it is recursing to the extend where it is too much. Thanks for pointing out. I just had performance issues working on a larger codebase using my version.

I'll switch regexes into iterative functions and do an another pull request. This should solve the issue, I believe